File: src/relu.zig
const std = @import("std");

pub fn Activation(comptime size: usize) type {
    return struct {
        last_inputs: []const f64 = undefined,
        fwd_out: []f64,
        bkw_out: []f64,
        batchSize: usize,

        const Self = @This();

        pub fn init(
            alloc: std.mem.Allocator,
            batchSize: usize,
        ) !Self {
            return Self{
                .last_inputs = try alloc.alloc(f64, size * batchSize),
                .fwd_out = try alloc.alloc(f64, size * batchSize),
                .bkw_out = try alloc.alloc(f64, size * batchSize),
                .batchSize = batchSize,
            };
        }

        pub fn forward(self: *Self, inputs: []f64) void {
            std.debug.assert(inputs.len == size * self.batchSize);

            var i: usize = 0;
            while (i < inputs.len) : (i += 1) {
                if (inputs[i] < 0) {
                    self.fwd_out[i] = 0.01 * inputs[i];
                } else {
                    self.fwd_out[i] = inputs[i];
                }
            }
            self.last_inputs = inputs;
        }

        pub fn backwards(self: *Self, grads: []f64) void {
            std.debug.assert(grads.len == size * self.batchSize);
            var i: usize = 0;
            while (i < self.last_inputs.len) : (i += 1) {
                if (self.last_inputs[i] < 0) {
                    self.bkw_out[i] = 0.01 * grads[i];
                } else {
                    self.bkw_out[i] = grads[i];
                }
            }
        }
    };
}

File: src/relu.java
import java.util.Arrays;

public class ReLU {
    private double[] lastInputs;
    private double[] fwdOut;
    private double[] bkwOut;
    private int batchSize;
    private int size;

    public ReLU(int batchSize, int size) {
        this.batchSize = batchSize;
        this.size = size;
        this.lastInputs = new double[size * batchSize];
        this.fwdOut = new double[size * batchSize];
        this.bkwOut = new double[size * batchSize];
    }

    public void forward(double[] inputs) {
        assert inputs.length == this.size * this.batchSize;

        for (int i = 0; i < inputs.length; i++) {
            if (inputs[i] < 0) {
                this.fwdOut[i] = 0.01 * inputs[i];
            } else {
                this.fwdOut[i] = inputs[i];
            }
        }
        this.lastInputs = Arrays.copyOf(inputs, inputs.length);
    }

    public void backwards(double[] grads) {
        assert grads.length == this.size * this.batchSize;

        for (int i = 0; i < this.lastInputs.length; i++) {
            if (this.lastInputs[i] < 0) {
                this.bkwOut[i] = 0.01 * grads[i];
            } else {
                this.bkwOut[i] = grads[i];
            }
        }
    }

    public double[] getFwdOut() {
        return fwdOut;
    }

    public double[] getBkwOut() {
        return bkwOut;
    }
}
